catboost/jvm-packages/catboost4j-prediction/src/main/java/ai/catboost/CatBoostModel.java

```diff
@@ -233,6 +234,36 @@ public static CatBoostModel loadModel(final InputStream in) throws CatBoostError
         return new CatBoostModel(handles[0]);
     }
 
+    /**
+     * Load CatBoost model from stream.
+     * You should consider this method for large files.
+     * @param in Input stream containing model.
+     * @return   CatBoost model.
+     * @throws CatBoostError When failed to load model.
+     * @throws IOException When failed to read model from file.
+     */
+    @NotNull
+    public static CatBoostModel loadModelUsingOffHeap(final InputStream in) throws CatBoostError, IOException {
+        int inputStreamSize = countAllBytes(in);
+        ByteBuffer memoryForModel = ByteBuffer.allocateDirect(inputStreamSize);
+        int bufferInBytes = 65536;
+        final long[] handles = new long[1];
+        ByteBuffer copyBuffer = ByteBuffer.allocateDirect(bufferInBytes);
+        byte[] copyBufferArray = copyBuffer.array();
+
+        while (in.read(copyBufferArray) != -1) {
+            memoryForModel.put(copyBufferArray);
+        }
+
+        implLibrary.catBoostLoadModelFromArray(memoryForModel.array(), handles);
+        return new CatBoostModel(handles[0]);
+    }
+
+    private static int countAllBytes(final InputStream in) throws IOException {
```

Comment:
It's not the best idea, because after this operation we have to start read stream from beginning. 
Not every stream support reset operation.